src/audio/google/google_rtc_audio_processing.c

```diff
@@ -822,23 +826,66 @@ static int google_rtc_audio_processing_process(struct processing_module *mod,
 	/* 32float: de-interlace ref buffer, convert it to float, skip channels if > Max
 	 * 16int: linearize buffer, skip channels if > Max
 	 */
+	/* Reduce cycle waste by streamlining the inner loop,
+	 * converting from array indexing to pointer arithmetic,
+	 * and putting data copy verification outside the loop.
+	 */
 	buffer_offset = 0;
-	for (int i = 0; i < cd->num_frames; i++) {
-		for (channel = 0; channel < cd->num_aec_reference_channels; ++channel) {
+	const int16_t *ref_end = ref + cd->num_frames * cd->num_aec_reference_channels;
+
+	/* Check if the calculated end of the reference buffer exceeds the actual end
+	 * of the buffer
+	 */
+	if (ref_end >= (const int16_t *)ref_buf_end)
+		/* If it does, wrap the reference buffer end pointer back to the start
+		 * of the buffer
+		 */
+		ref_end = (const int16_t *)ref_buf_start;
```

Comment:
Would something like ref_data_end be a better name? I was first confusing this with ref_buf_end.